### ISSUE TITLE
Install "Custom CA Trust" on AKS nodes

### ISSUE THREAD

**Issue by jabbera** [Azure/AKS#2259]
Private clusters cause AKS internet traffic to be directed out customers internet edge devices. We have a decrypting firewall. Therefore we need our Certificate Authority root cert installed on every node in our cluster or every package mirrored to our local repo as to not be intercepted by our MITM firewall. I was originally going the mirroring route until coredns broke last night during what I assume to be an upgrade, took down the whole cluster, and I didn't know how to update the image location. (Is this possible in AKS?) 

Edit: Actually this is WAY worse for a new cluster. The deployment failed and I had to go in and dig around to find this:

![image](https://user-images.githubusercontent.com/301208/114317623-f0e2b400-9ad6-11eb-9e9c-44bb0ffea7dd.png)

Now I have a bootstrapping issue. I can't deploy a cluster without my CA installed but my automation that installs my CA doesn't run till after the cluster is provisioned. Luckily I was able to apply the CA installation daemonset described below, re-run my deployment and my cluster ended up in an alright state. 

Customizing the CA repo should be built in node customization functionality to private clusters instead I had to find and follow these directions as well as augment them to be sure they run on every node: http://hypernephelist.com/2021/03/23/kubernetes-containerd-certificate.html

Please make this core functionality!

**Comment by miwithro** [comment]
Thanks for the request.  Putting this on our backlog.

**Comment by alxzndr** [comment]
We face the same issue since to move to containerd.

On previous versions with docker we were able to configure our registry ca trough a daemonset.
This daemonset puts ca config files for our container registry in the /etc/docker/certs.d folder.
Docker automatically picked up these cert files.

Containerd has a similar functionality but this requires the registry configpath parameter to be set.
By default this parameter is empty on the AKS worker nodes. 

If this registry config_path parameter is set to a default value we would not need daemonsets with elevated rights to be able to restart containerd. 
```
  [plugins."io.containerd.grpc.v1.cri".registry]
    config_path = "/etc/containerd/certs.d"
```

**Comment by jabbera** [comment]
Just to add this SIGNIFICANTLY increases node startup time. The daemon set has to run then all the core images need to be pulled. I'm averaging 8 minutes for node startup. It's sad:-(

**Comment by AndreasMWalter** [comment]
@jabbera  
We were just evaluating another workaround, which at preset seems to work however it is really... uhm... botched:
With the AKS Proxy Feature enabled we have created a dummy proxy entry and noProxy "*" and injected an onpremises Trusted Root CA. That seems to work, however obviously the AKS Proxy Feature is still in preview and this workaround is also probably not supported by Microsoft. (And might cause other issues)
```
{
    "httpProxy": "http://dummy.proxy/",
    "httpsProxy": "https://dummy.proxy/",
    "noProxy": ["*"],
    "trustedCa": "BASE64ENCODEDROOTCA"
}
```

**Comment by jabbera** [comment]
@AndreasMWalter this is genius. I'd love an official statement from MSFT about it.

**Comment by AndreasMWalter** [comment]
> 
Yes we are still evaluating this, as I said some things may be wonky for example:
Due to the fact of how "NO_PROXY" works and Microsofts implementation adds automatically certain values: 
```
169.254.169.254,10.0.0.0/16,AKSPRIVATELINKHERE.privatelink.westeurope.azmk8s.io,10.65.8.0/25,10.244.0.0/16,172.17.0.1/16,168.63.129.16,konnectivity
``` 

However if `curl` for example finds anything after "*" it will fail.
we are currently considering injecting an empty string as value for httpProxy and httpsProxy, however the API might block us from doing that so well...

Or injecting all existing Top Level Domains ->
https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy/

**Comment by niklev** [comment]
this is very valuable for enterprise use cases

**Comment by miwithro** [comment]
@alexeldeib

**Comment by jabbera** [comment]
Sorry to be a pain but I'm rolling out 4 new clusters and I hitting this left right and center. Any idea when it's coming out?

**Comment by alexeldeib** [comment]
we're trying to finalize an approach to work on it in feb. the current preference is for AKS to take an array of Keyvault secret IDs which contain the desired certificates, and nodes would periodically refresh them from KV (which covers update). We took the raw cert for http proxy but that makes updates more disruptive than necessary as you currently need to roll the nodes for new certs to get picked up.

**Comment by jabbera** [comment]
Keyvault is great as long as we can use a Private Endpoint only or AKS is trusted. In my experience the downside to taking on another service as a dependency is it has to work with public access disabled for it to be useful in my space. This is especially applicable in this case as we are hosting aks as a private cluster. Attaching them to the arm/bicep template would be 100x easier but whatever approach works as long as it's all actually private.

**Comment by alexeldeib** [comment]
> Attaching them to the arm/bicep template would be 100x easier but whatever approach works as long as it's all actually private

this unfortunately results in other issues for us, mostly around the VMSS update flow + custom data size limits (latter particularly for windows, although that may not be super relevant to you). 

> or AKS is trusted

can you clarify that piece? not sure I followed.

> as long as we can use a Private Endpoint only

for clarity, something inside the cluster hitting private endpoint _before_ CA is setup is fine? we've played around with a few different approaches between raw certs, KV references, as well as where the certs are resolved -- passed raw from user to custom data, resolved from KV inside RP and passed to custom data that way, passed as KV IDs to node which uses kubelet identity to resolve secrets and download to node (current direction).

we've been around the bush with this one internally quite a bit, the strong current preference is basically a managed version of the daemonset everyone ends up running anyway for this task. we'd precache a docker image with a well-known tag and use that in the DS. this avoids the egress issues you shared in your screenshot pulling images from MCR through a TLS decrypted FW before the cert is configured. it also cleanly handles update scenarios while also working well for windows (albeit with some constraints). 

doing the private endpoint in-cluster is presumably fine (? asking) but private endpoint presumably doesn't work if we resolve the certs from anywhere outside the subnet/vnet as you indicated.

**Comment by jabbera** [comment]
I think we are on the same page. I'll repeat what I think you said:

We would point our AKS cluster to a KeyVault that could host our root/intermediates. A managed daemonset would reach out and pull these certs at startup and could be done over a KeyVault Private Endpoint.

(When I said trusted I meant a trusted service so I didn't have to have a public ip added to KeyVault. Private endpoint is better so if the above scenario works we are all set.)

**Comment by kekovski** [comment]
What is the status of this feature? Is there any ETA?

**Comment by miwithro** [comment]
@kekovski End of March will have a Public Preview.

**Comment by miwithro** [comment]
@jabbera can you give https://docs.microsoft.com/en-us/azure/aks/custom-certificate-authority to see if that works for your scenario?

**Comment by mezzofix** [comment]
Hi, got a full private AKS cluster with egress routed via the firewalls in hub subscription by setting the outboundType to value of userDefinedRouting. Getting now `Warning  Failed     4s (x2 over 17s)  kubelet            Failed to pull image "nginx:1.14.2": rpc error: code = Unknown desc = failed to pull and unpack image "docker.io/library/nginx:1.14.2": failed to resolve reference "docker.io/library/nginx:1.14.2": failed to do request: Head "https://registry-1.docker.io/v2/library/nginx/manifests/1.14.2": x509: certificate signed by unknown authority` Will look at the solution above (https://docs.microsoft.com/en-us/azure/aks/custom-certificate-authority) but there's something I don't understand. Since I'm using an existing VNet/subnet for AKS and a UDR was applied to point 0.0.0.0/0 to the firewall in hub, why did the AKS cluster provisioning not fail due to certificate verification as it does on the worker nodes ...

**Comment by jabbera** [comment]
@miwithro sorry. I was down hard the last few weeks. This looks okay as long as the cluster won't fail to provision. I have a bicep template that will deploy the cluster/nodepools, then I would add/update the secret in the next step of the pipeline. If the deployment fails because of certificate issues my next step would not run and I'm back to having a non-ideal DR deployment plan.

**Comment by miwithro** [comment]
@jabbera great.  Looking forward to any feedback

**Comment by mkemmerz** [comment]
@miwithro Thanks for linking the documentation. I have a question as the Azure CLI blocks me at the moment.

- Which Azure CLI Version supports the ` --enable-custom-ca-trust` flag? I have updated to the [newest 2.37](https://github.com/Azure/azure-cli/releases) of Azure CLI and the [release notes](https://docs.microsoft.com/en-us/cli/azure/release-notes-azure-cli) doesn't mention anything to this flag.

![image](https://user-images.githubusercontent.com/46575727/173504727-30697268-381d-4483-9909-c8426b1cf737.png)

**Comment by mkemmerz** [comment]
Okay so I tried to use the feature but wasn't successfully. Maybe someone can confirm that the issue is on my end?
I was trying using 6 certificates and my use-case was to pull the pod images from an external service like Nexus. Pods were stuck in imagepullbackoff with certificate error.

```
apiVersion: v1
kind: Secret
metadata: 
    name: custom-ca-trust-secret
    namespace: kube-system
type: Opaque
data:
    ca1.crt: |
      xx=
    ca2.crt: |
      xx==
    ca3.crt: |
      xx==
    ca4.crt: |
      xx
    ca5.crt: |
      xx==
    ca6.crt: |
      xx
```

**Comment by protonyx** [comment]
Same question as @mkemmerz, we installed the feature and enabled it on the node pools but am still getting the certificate errors. Our certificates are PEM formatted.

I didn't see any evidence in the nodes that the certificates are being loaded, nor are there any troubleshooting steps in the documentation. Is there a way to get logs or otherwise tell that the certificate loader/watcher is running?

**Comment by Eldarrin** [comment]
Same issue as @mkemmerz & @protonyx. Is there somewhere that it logs to so I can deep dive the diagnosis. Using 2 certs and 4 nodes all in systempool in a private cluster

**Comment by UtheMan** [comment]
Hi, 

@mkemmerz  - You can add more certificates, we just used 2 as an example. Will make sure to have the docs updated to make it clear.

@mkemmerz @Eldarrin @protonyx - did you enable the feature on an existing node pool? We identified an issue with the feature, where enabling it on an existing node pool doesn't work as expected. A fix is on the way (will post here once it is released). To see if the feature is working as expected for you, could you try creating a new cluster with the feature enabled? 

Once the fix is live we will also update docs with instructions on how to troubleshoot.

**Comment by protonyx** [comment]
I attempted to create a new node pool with the `--enable-custom-ca-trust` flag, but got a very unhelpful error:

```
az aks nodepool add \
    --resource-group my-resource-group \
    --cluster-name my-cluster-name \
    --name catestpool \
    --zones 1 2 3 \
    --node-vm-size Standard_D16as_v4 \
    --node-osdisk-type Ephemeral \
    --node-osdisk-size 128 \
    --os-type Linux \
    --node-count 1 \
    --mode User \
    --enable-custom-ca-trust
The behavior of this command has been altered by the following extension: aks-preview
The behavior of this command has been altered by the following extension: aks-preview
(InternalOperationError)
Code: InternalOperationError
Message:
```

The provisioning state for the new node pool is listed as `Failed` and 0 nodes are actually running, despite it saying `count=1`:
```
{
    "availabilityZones": [
      "1",
      "2",
      "3"
    ],
    "count": 1,
    "currentOrchestratorVersion": "1.22.4",
    "enableCustomCaTrust": true,
    "mode": "User",
    "name": "catestpool",
    "nodeImageVersion": "AKSUbuntu-1804gen2containerd-2022.06.22",
    "orchestratorVersion": "1.22.4",
    "osDiskSizeGb": 128,
    "osDiskType": "Ephemeral",
    "osSku": "Ubuntu",
    "osType": "Linux",
    "powerState": {
      "code": "Running"
    },
    "provisioningState": "Failed",
    "vmSize": "Standard_D16as_v4",
    ...
  }
```

I did see that a daemonset called `custom-ca-trust` was created on the cluster in the `kube-system` namespace, but there are 0 pods despite having a nodeAffinity that includes 2 of the 4 node pools in our cluster (which is strange, since I enabled the feature on all of the node pools). The container image tag referenced by the daemonset doesn't seem to exist (`mcr.microsoft.com/aks/aks-node-ca-watcher:static`), but according to the vhd nodes for the image the nodes are running, that image should be pre-pulled with a different tag: `mcr.microsoft.com/aks/aks-node-ca-watcher:master.220422.1` (https://github.com/Azure/AKS/blob/master/vhd-notes/aks-ubuntu/AKSUbuntu-1804/2022.06.22.txt#L172)

Considering the headache required to get a new cluster provisioned in my company's environment, I will hold off on testing if the feature works on a fresh cluster.

**Comment by UtheMan** [comment]
@protonyx thank you for checking that, looking into it.

**Comment by MattGarner-N** [comment]
Is there a plan for this feature on Windows Nodes? I can't find any way to install custom CA certs on windows nodes...

**Comment by AbelHu** [comment]
@MattGarner-N did you try to use windows host process containers (HPC) to install custom CA on Window nodes? HPC is supported in k8s 1.23+ and containerd 1.6+

**Comment by MattGarner-N** [comment]
Hi @AbelHu, I did try but cannot find a way... No documentation on it either as Windows Node Pools are not widely used.

Please let me know if you have a solution! :)

**Comment by AbelHu** [comment]
@MattGarner-N please update the PowerShell command and try to deploy below HPC pod in your Windows nodes
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: hpc
  labels:
    app: hpc
spec:
  selector:
    matchLabels:
      app: hpc
  replicas: 1
  template:
    metadata:
      labels:
        app: hpc
    spec:
      securityContext:
        windowsOptions:
          hostProcess: true
          runAsUserName: "NT AUTHORITY\\SYSTEM"
      hostNetwork: true
      containers:
      - name: hpc
        image: mcr.microsoft.com/windows/servercore:1809
        args:
        - powershell.exe
        - -Command
        - "[REPLACE THIS WITH YOUR COMMAND TO INSTALL CUSTOM CA IN THE HOST]"
        imagePullPolicy: IfNotPresent
      nodeSelector:
        kubernetes.io/os: windows
```

**Comment by megakid** [comment]
I've tried this and it does not install the certificate onto the host.  I tried:

`& certutil -addstore -enterprise -f -v root C:\certs\root_ca.crt` => imported successfully but a separate image pull to our internal docker registry with self-signed certificate still reported x509 certificate error.

and

`Import-Certificate -FilePath "C:\certs\root_ca.crt" -CertStoreLocation Cert:\LocalMachine\Root` => `Import-Certificate : Error HRESULT E_FAIL has been returned from a call to a COM component.`

**Comment by MattGarner-N** [comment]
@AbelHu  FYI see above

**Comment by AbelHu** [comment]
> I've tried this and it does not install the certificate onto the host. I tried:
> 
> `& certutil -addstore -enterprise -f -v root C:\certs\root_ca.crt` => imported successfully but a separate image pull to our internal docker registry with self-signed certificate still reported x509 certificate error.

@megakid Did you check whether the imported cert exists in the host? Could you manually pull your image in the host after confirming that the cert exists? 

> and
> 
> `Import-Certificate -FilePath "C:\certs\root_ca.crt" -CertStoreLocation Cert:\LocalMachine\Root` => `Import-Certificate : Error HRESULT E_FAIL has been returned from a call to a COM component.`

Can you run this command in the host successfully? (You can use a different node for the test)

**Comment by cgreenza** [comment]
Hi, enabling as part of new cluster creation seems to work fine, but enabling on existing cluster/node pool does not (`custom-ca-trust` daemon set not created). In July @UtheMan mentioned that a fix for this is on the way. Any ETA for fix or temporary workaround? Thanks

**Comment by cgreenza** [comment]
Pods for `custom-ca-trust` daemon set are not scheduled on system nodes with `CriticalAddonsOnly=true:NoSchedule` taint. Please consider adding a toleration for this.

**Comment by alexeldeib** [comment]
ack, @utheman fyi.

**Comment by djniu** [comment]
one question, may this feature affects the pods in nodes? Do we need to trust again in pods if we need to trust custom CA in pods?

**Comment by EvertonSA** [comment]
+1 for this feature to go GA

**Comment by EvertonSA** [comment]
colleagues, on the docs https://learn.microsoft.com/en-us/azure/aks/custom-certificate-authority we have:
- Configure a new AKS cluster to use a custom CA
- Configure a new nodepool to use a custom CA
- Configure an existing nodepool to use a custom CA

I dont see something like "Configure an existing AKS cluster to use a custom CA". Is this intended? I used the procedure on "Configure an existing nodepool to use a custom CA" but no daemonset was installed in my cluster.

**Comment by EvertonSA** [comment]
Hi colleagues, any feedback on my comments above? My boss is on my back and I don't want to use this http://hypernephelist.com/2021/03/23/kubernetes-containerd-certificate.html if Microsoft has an official way of doing it

**Comment by miwithro** [comment]
@EvertonSA, we are working on an official way (Early December).

**Comment by UtheMan** [comment]
Hi @EvertonSA, will be taking a look at your issue this week. Apologies for the delay - i was out of the office.

**Comment by UtheMan** [comment]
@EvertonSA could you give some more details about your exact setup/situation?

Due to current implementation there is some amount of time (10+ minutes sometimes) you need to wait for the daemonset to come up. Could you confirm if its still not visible? 

Which node image version are you using?

**Comment by EvertonSA** [comment]
hi @UtheMan, thanks for the message. I can confirm the daemonset was installed, its named `
custom-ca-trust` and it resides on `kube-system` namespace. Now the problem is that the daemonset has no tolerations, thus, cannot be placed in the nodepool, problem that was already mentioned here https://github.com/Azure/AKS/issues/2259#issuecomment-1253794956

**Comment by EvertonSA** [comment]
just for reference, we upgraded the kubernetes version on existing node pools and the flag was disabled. I assume whenever this https://github.com/hashicorp/terraform-provider-azurerm/issues/18391 is implemented we will not face it again. 

also, I noticed v20221030 is available on Europe, so I'm giving a shot. Will post my results here

**Comment by EvertonSA** [comment]
ok, some updates from me. I have 4 clusters, but this is a must on only 2 clusters in a specific node pool. 

I ran the ` az aks nodepool update --resource-group RG --cluster-name CLUSTER  --name NODEPOOL --enable-custom-ca-trust` on both at the same exact time.

the second cluster took about 1 hours to have the daemonset. the first one took over a day (28 hours to be precise). 

at the end, feature wise, LGTM. but this insane delay on installing the daemonset is absurd. We should add to documentation that it can take up to 24 hours to have your daemonset. I found no way of hard triggering this.

**Comment by UtheMan** [comment]
Hi @EvertonSA, thanks for checking!
This definitely shouldn't be taking that much time. I'll be looking into this today and will let you know about my findings.

**Comment by EvertonSA** [comment]
hi @UtheMan, thanks that really helps. this feature is working very nice, and I would recommend to all customers that have the need for it. I'm eager to have this on terraform. thanks.

**Comment by UtheMan** [comment]
Thanks @EvertonSA!  I'm working on a fix to have the DS come up faster (10-15 mins at most) without running the extra command.
After Thanksgiving i'll be getting in touch with folks to get it working on Terraform as well.

**Comment by marrotte** [comment]
Happy Holidays @UtheMan :)  So glad to find this issue and conversation.  Any updates the on the DS timing and terraform?  We're hitting a similar issue with the latest recommend node image version AKSUbuntu-1804gen2containerd-2022.11.02.  Any updates are greatly appreciated.  

We also observer a couple more things like in the DS logs `2022/12/06 05:58:59 Couldn't read file ..2022_12_06_05_58_58.3057698138, error: read opt/certs/legacy/..2022_12_06_05_58_58.3057698138: is a directory
2022/12/06 05:58:59 Couldn't read file ..data, error: read opt/certs/legacy/..data: is a directory` and in addition to the `custom-ca-trust-secret` , we're seeing an empty `custom-ca-trust-secret-generated` secret appear.

**Comment by UtheMan** [comment]
Hi @marrotte, happy Holidays!

Improvements to DS timing should be released along with v20221204 version, during my tests it seemed to be working as expected. Please let me know if that won't be the case for you and I'll be happy to investigate more. In the meantime you can continue to use the command I shared: `az aks update -g resource-group-name --name cluster-name` to trigger update which will make the DS appear. 

The error about directories you see are expected at the moment, this is caused by the way that binary deployed in custom ca trust pod is checking for new certs. I will look into refactoring that to avoid unnecessary noise. 


Empty `custom-ca-trust-secret-generated` is related to upcoming changes to the feature (we will add an option to specify CAs that will be added to trust store during node boot up process, allowing users that need their CAs to e.g. access private registries during provisioning), I will link docs/more details here once the changes are released (most likely later this week/early next week).

**Comment by dhirschfeld** [comment]
> *When it comes to Terraform changes - I am currently looking into adding custom-ca related fields to it, that would include both currently existing --enable-custom-ca-trust and the field related to new changes I mentioned above. I cannot really provide an ETA for that atm, I will try to make PRs before Christmas time. Will update here once I have a concrete ETA/PR ready.*

Has this not been resolved by https://github.com/hashicorp/terraform-provider-azurerm/pull/19546?

**Comment by UtheMan** [comment]
Thanks for linking that @dhirschfeld, I wasn't aware someone else has already done it. Had this in my backlog for the upcoming weeks...

**Comment by UtheMan** [comment]
Hi all, apologies for double comment. We added some new things to the feature, you can read more about that in the [docs](https://learn.microsoft.com/en-us/azure/aks/custom-certificate-authority).

In the new year there will be more changes to the feature as we'll be preparing for this to go to GA.

**Comment by EvertonSA** [comment]
I'm not sure when this started, but the DS shows this logs:

```
2023/01/13 08:46:41 Couldn't read file ..2022_12_02_10_32_04.1365194661, error: read opt/certs/generated/..2022_12_02_10_32_04.1365194661: is a directory
2023/01/13 08:46:41 Couldn't read file ..data, error: read opt/certs/generated/..data: is a directory
2023/01/17 13:56:01 Couldn't read file ..2023_01_17_13_55_57.2857493483, error: read opt/certs/legacy/..2023_01_17_13_55_57.2857493483: is a directory
2023/01/17 13:56:01 Couldn't read file ..data, error: read opt/certs/legacy/..data: is a directory
2023/01/17 13:56:01 Couldn't read file ..2022_12_02_10_32_04.1365194661, error: read opt/certs/generated/..2022_12_02_10_32_04.1365194661: is a directory
2023/01/17 13:56:01 Couldn't read file ..data, error: read opt/certs/generated/..data: is a directory
```

**Comment by UtheMan** [comment]
Hi @EvertonSA 

1. Those logs are expected, I haven't had a chance to silence directory related errors like this yet. We are currently gathering some telemetry for the feature and once that is done I will be circling back and adding required changes. I will make sure to improve logging then
2. Yes unfortunately in some cases containerd restart might be required for this to work. We have more details in [docs](https://learn.microsoft.com/en-us/azure/aks/custom-certificate-authority). You could try using the approach with certs passed during node provisioning, that will not require a restart, but there are some other tradeoffs (see docs for more info) 😄. 
3. As for the feature registration, this feature is currently in preview and because of the preview feature policy this registration will be required until this feature is brought to general availability (AKS preview features are available on a self-service, opt-in basis. Previews are provided "as is" and "as available," and they're excluded from the service-level agreements and limited warranty. AKS previews are partially covered by customer support on a best-effort basis. As such, these features aren't meant for production use). As I mentioned we are gathering more data and once that is done we will make the final push to bring this to GA.

**Comment by shikhasinner** [comment]
Do we have an ETA when is this feature likely to go GA ? A lot of FSI customers in ANZ are asking for this.

**Comment by anwarchk** [comment]
Is there a terraform or cluster API equivalent of this feature coming anytime soon ? There is a feature flag to enable the custom ca cert, but not a way to provide certificate details inside terraform. We still have to rely on Az CLI to enable this, which is not scalable when we are support 50 plus development teams with multiple clusters.

**Comment by jrwhite17** [comment]
Are there any updates on this issue? How are other Azure customers working around this issue?
Also, does anyone have a list of all the public endpoints AKS needs to interact with? So far I'm tracking:
mcr.microsoft.com:443
*.data.mcr.microsoft.com:443
*.opinsights.azure.com:443

**Comment by alexeldeib** [comment]
> list of all the public endpoints AKS needs

Take a look at this: https://learn.microsoft.com/en-us/azure/aks/limit-egress-traffic

**Comment by EvertonSA** [comment]
> Is there a terraform or cluster API equivalent of this feature coming anytime soon ? There is a feature flag to enable the custom ca cert, but not a way to provide certificate details inside terraform. We still have to rely on Az CLI to enable this, which is not scalable when we are support 50 plus development teams with multiple clusters.

I enabled the [custom_ca_trust_enabled](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/kubernetes_cluster#custom_ca_trust_enabled) then I have in the same terraform created a secret with correct naming and works fine.

the only thing that terraform cannot automate is enabling Microsoft.ContainerService/CustomCATrustPreview

**Comment by allyford** [comment]
Update: GA is in progress. Expected GA by end of 2024.

**Comment by allyford** [comment]
Update: New GA target is Feb 2025.

**Comment by allyford** [comment]
Update: GA on track for Feb 2025.
Note: This is an estimate and is subject to change.

**Comment by allyford** [comment]
[Custom Certificate Authority](https://learn.microsoft.com/azure/aks/custom-certificate-authority) allows you to create custom CAs and apply them to your cluster. This feature is targeting GA in March 2025 with the ability to install CAs during cluster creation using the [az aks create](https://learn.microsoft.com/cli/azure/aks#az-aks-create) command and specifying your text file for the --custom-ca-trust-certificates parameter.

The following use cases will not be in scope for this release:

1. Install CAs after cluster creation 
2. Private Registry Authentication

**Comment by allyford** [comment]
Update: GA targeting end of March/early April.
Note: This is an estimate and is subject to change.

**Comment by allyford** [comment]
Custom Certificate Authority is now available in the GA CLI in v2.72.0
